invenio_drafts_resources/services/records/service.py

```diff
@@ -0,0 +1,228 @@
+            record_cls=self.config.record_cls,
+            record_to_index=self.config.record_to_index
+        )
+
+    # Draft attrs
+    @property
+    def draft_cls(self):
+        """Factory for creating a record class."""
+        return self.config.draft_cls
+
+    # High-level API
+    # Inherits record read, search, create, delete and update
+
+    def read_draft(self, id_, identity, links_config=None):
+        """Retrieve a draft."""
+        # Resolve and require permission
+        # TODO must handle delete records and tombstone pages
+        # TODO: Can we make in a similar fashion than "create"?
+        draft = self.draft_cls.pid.resolve(id_, registered_only=False)
+        self.require_permission(identity, "read_draft", record=draft)
+
+        # Run components
+        for component in self.components:
+            if hasattr(component, 'read_draft'):
+                component.read_draft(identity, draft=draft)
+
+        return self.result_item(
+            self, identity, draft, links_config=links_config)
+
+    def update_draft(self, id_, identity, data, links_config=None):
+        """Replace a draft."""
+        # TODO: etag and versioning
+        draft = self.draft_cls.pid.resolve(id_, registered_only=False)
+        # Permissions
+        self.require_permission(identity, "update_draft", record=draft)
+        data, _ = self.schema.load(
+            identity, data, pid=draft.pid, record=draft)
+
+        # Run components
+        for component in self.components:
+            if hasattr(component, 'update_draft'):
+                component.update_draft(
+                    identity, record=draft, data=data)
+
+        # TODO: remove next two lines. (See also record-resources)
+        draft.update(data)
+        draft.clear_none()
+        draft.commit()
+
+        if self.indexer:
+            # FIXME: This would double index. Need delete->index?
+            self.indexer.index(draft)
+
+        return self.result_item(
+            self, identity, draft, links_config=links_config)
+
+    def create(self, identity, data, links_config=None):
+        """Create a draft for a new record.
+
+        It does not eagerly create the associated record.
+        """
+        # FIXME: This won't work with partial validation
+        return self._create(
+           self.draft_cls, identity, data, links_config=links_config
+        )
+
+    def _patch_data(self, record, data):
+        """Temporarily here until the merge strategy is set."""
+        # TODO: Implement better update/merge strategy
+        for key in data.keys():
+            record[key] = data[key]
+
+    def edit(self, id_, identity, links_config=None):
+        """Create a new revision or a draft for an existing record.
+
+        Note: Because of the workflow, this method does not accept data.
+        :param id_: record PID value.
+        """
+        self.require_permission(identity, "create")
+        # Get draft
+        try:  # Draft exists
```

Comment:
minor: maybe just extract this logic in a `_edit_or_create` method. It will make the code easier to read :)

Reply:
That is how it was before (what we debugged yesterday, but we need to know the `record_cls` and `draft_cls`). So:

a) We make the Draft class know which class it is drafting.
b) We can extract that function but at service level

Reply:
Sorry I meant the `b`. Extract that function in the service level ;) I am not sure if is reusable, but I think it would make the code a bit more readable...No strong opinion though :)

Reply:
Agreed, done :)